### PR TITLE
WIP: Cherow 2

### PR DESCRIPTION
**Pull request by KFlash**
@cherow/cherow-team 

This is a WIP PR for the upcoming version 2.0 of Cherow. It's still in early stage, but it now works.

**Note:** Many things have changed since 1.6.x, and some new options have been added. Please look at the issues for further updates on this.

The AST will not be backward compatible and not compatible with other parsers due to the fact that Cherow 2.0 now strictly follows the ESTree specifications. 

**Be aware** that Cherow 2.0 also strictly follows the ES specifications and let you parse with and without "AnnexB". If you choose to turn off the web compability, 2.0 will act differently.

**TODO:**

- [x] Refactor scanner 

- [ ] Implement regex validator

- [x] Introduce new options and make sure they work

- [x] Implement lexical and block scoping

- [x] Re-develop statement code and re-use if possible from 1.6.8

- [x] Re-develop expression parsing

- [ ] Optimize for performance

### CODE REVIEW

**Comment by g-plane** [comment]
What a huge PR!

**Comment by KFlash** [comment]
Simply because I only put a placeholder. Everything will be fixed

**Comment by KFlash** [comment]
@fkleuver if you look at what I have pushed you will notice that the lexer is faster now. I also removed and simplified it. Also lookahead can be done easy. I can still optimize a few things.

Why is circleci failing?

**Comment by fkleuver** [comment]
CircleCI isn't failing. travis and appveyor are because I removed their configs, you should disable them (I can't access those)

**Comment by KFlash** [comment]
Why I get this?

**Comment by KFlash** [comment]
@fkleuver you are admin, so you can disable in settings?

**Comment by KFlash** [comment]
@fkleuver @g-plane @Aladdin-ADD Too avoid too many questions. If you working on a parser, you should also know the ECMA specs. I say this in case you guys figure out that the HTML comments etc doesn't work anymore as in other parsers. Reason? See AnnexB. HTML comments are only supported if AnnexB is enabled. 

I'm doing both sides now - with and without AnnexB. There are other things that have changed now too.

**Comment by KFlash** [comment]
@fkleuver How to lint this code? Can this be automatically done? And it doesn't seem to allign properly

**Comment by KFlash** [comment]
@Aladdin-ADD You are used to collect the comments in the AST output? Not a callback?

I fixed comment attachment now. @fkleuver For now I attached the comment on the AST output. Should we use a callback ala Acorn for instead ?

**Comment by Aladdin-ADD** [comment]
the new options LGTM!

should we rename "module": true|false => "sourceType": "script"|"module"?

**Comment by KFlash** [comment]
@Aladdin-ADD It's possible, but will add extra code for a now simple line of code. It's only purpose is to run in module code if set to true. E.g. `sourceType: "srcipt"` have no effect at all.

**Comment by Aladdin-ADD** [comment]
I found the "module" option has been included in v1, so it might not worthy to changing for less confusing.

**Comment by KFlash** [comment]
@Aladdin-ADD Can you try to lint this PR? I have some RL stuff to fix for now

**Comment by Aladdin-ADD** [comment]
sure, I will try.

**Comment by g-plane** [comment]
Can we use Prettier to avoid this?

**Comment by KFlash** [comment]
@fkleuver  @g-plane @Aladdin-ADD I have put together a basic lexer now and it will also parse and output a "TODO!".  Note that things are little more different than before.

@fkleuver Can you modify **token.ts**

**Comment by KFlash** [comment]
I'm adding *"scooping"* to Cherow now, and this will be refactored for performance later. There will soon be optional to use it. If you use the `native` option this will have no effect. I'm passing the scope as arguments where it's needed due to performance reasons.

**Comment by KFlash** [comment]
@fkleuver @Aladdin-ADD @g-plane My upcoming changes to Cherow will probably *"blow your head off"* so please don't try to follow me anymore or ask to much questions. My first priority is the performance as we all know.

**Comment by KFlash** [comment]
@fkleuver @g-plane @Aladdin-ADD 

Ok, guys. It looks like Cherow finally is coming back!

This `const Cherow = IS_BACK!!` outputs:

```js
{
        "body": [
          {
            "declarations": [
              {
                "id": {
                  "name": "Cherow",
                  "type": "Identifier",
                },
                "init": {
                  "name": "IS_BACK",
                  "type": "Identifier",
                },
                "type": "VariableDeclarator",
              }
            ],
            "kind": "const",
            "type": "VariableDeclaration",
          },
        ],
        "sourceType": "script",
        "type": "Program",
      }
```
**Please note!!** I haven't checked the performance yet, and behind the scene there are scoping activated included with lexical binding checks. This means that e.g. this one `const a = b, a = c` is invalid.

This is all new in version 2.0, and I added it because Acorn have a crapy scope implementation. This can be disabled if you set the `native` option

**Comment by KFlash** [comment]
I have implemented lexical bindings / scoping. Working on block scoping and function args. 
This turns out to be very tricky. I think that's why acorn doesn't support block scoping...

**Comment by KFlash** [comment]
@fkleuver Read your git

**Comment by KFlash** [comment]
@g-plane @Aladdin-ADD @fkleuver Have a look at the new tests that I added to get a hint on many of the new changes in Cherow 2.0.

**Comment by KFlash** [comment]
@g-plane @g-plane @fkleuver I have now implemented complete block scoping and lexical bindings into Cherow. Acorn come to short!!  

Is this something we will let the user turn off or not through options?

**Comment by KFlash** [comment]
@fkleuver Can you update token.ts with new precedence tokens as we talked about before? Kind of needed because I'm going to do a infinite state machine trick on it

**Comment by g-plane** [comment]
So Cherow can produce not only AST but also scope analysis?

**Comment by KFlash** [comment]
The scope is block scoping if you read the specs. So outputs only ast and
soon you can turn off ast creation to for only  validation in e.g ide or
text editor

**Comment by KFlash** [comment]
@fkleuver For now I just get things working, but I do a couple of tricks now so we only need to pare a simpler, faster sub-grammar if it's evident instead of calling a lot of sub-functions to archive the same goal.

I'm now using nested functions to get it working, but soon as it does I may need to borrow your brain if it's okay for you to flatten the whole shit into a single function with only a precedence value as 3rd arg. This require modification of token.ts.

Are you with me on this ?

Here you can see an example on how I do things:

```js
function parseLiteralExpression(state: ParserState, context: Context) {
    const expr = parseAssignmentExpression1(state, context, parseMemberExpression(state, context, parseLiteral(state, context)));
    if (state.currentToken !== Token.Comma) return parseExpressionStatement(state, context, expr);
    return parseExpressionStatement(state, context, parseSequenceExpression(state, context, expr));
}
```

**Comment by KFlash** [comment]
@fkleuver @Aladdin-ADD @g-plane  
Guys! Sorry for not publishing anything now, but Cherow 2.0 turns out to be a PITA and very complex. I think it's best that I actually get things working before I push anything. Agree?

Current state of Cherow 2.0 on my locale computer is that Cherow now is working again. I brought it back to life and it can do simple parsing without too much validations. Statements, expressions and declarations now all works again!

But ATM everything is a mess, not optimized and way too many bitmasks and function args to keep track on. Plus the code isn't up to my standards.

However. Expect this things working in my next push

- [ ] Statement parsing

- [ ] Block scoping

- [ ] Lexical bindings

- [ ] Expression parsing

- [ ] Argument validations

- [ ] AnnexB fiddling

No module code yet!

**Comment by KFlash** [comment]
@fkleuver One of the *big* issues in 1.6.8 was that we forgot to implement duplicate arg check for arrow functions, and there was no lexical binding and block scope involved.

Look at [this](https://github.com/cherow/cherow/blob/532e729f7eb422710fec7c30b288df174683517e/test/parser/expressions/arrows.ts) tests. It's all fixed.

**Comment by KFlash** [comment]
@cherow/cherow-team Just to let you all know that Cherow 2.0 now is alive and can do simple parsing, and it far from what v.1.6x was. In fact very different in the way it behave. 

In a few days from now I will push some code that can be used for parsing, but almost zero validation. The only validation that exist for now is scoping and lexical bindings. Please don't expect miracles. Things are still early **wip**.

@fkleuver You think you can find a place in your busy schedule to help out and do some Cherow code? I may need a second brain to implement something in a most performant way.

**Comment by KFlash** [comment]
@cherow/cherow-team I just dumped my latest tests. Take a look at them to get an idea on what v. 2.0 have become!

@fkleuver What do you think about [this test cases](https://github.com/cherow/cherow/blob/72a0060a9383fe57e3d532d766dcd835beab01f0/test/parser/expressions/object.ts#L24)? They are not failing in v. 1.6.x

**Comment by fkleuver** [comment]
> @fkleuver What do you think about this test cases? They are not failing in v. 1.6.x

I see my name in there :P
Btw, no location tracking verification yet?
Looks like a lot of work to write tests that way! Why not more looping to generate em?

> @fkleuver You think you can find a place in your busy schedule to help out and do some Cherow code? I may need a second brain to implement something in a most performant way.

Well I've attempted this in the past and it resulted in quite a bit of backlash from you if you remember. Strange/vague complaints, doubts, lots of discussions, etc.

If you can just make sure it's spec compliant and there are sufficient tests to validate that, I would be more than happy to help out polishing and optimizing things however.

**Comment by KFlash** [comment]
No location tracking yet because trying to come up with a better solution...

Regarding past. As it is today I don't care much how the code works atm
only the fact that it works. If u seen the source at current stage you
would think I have a mental illness. So if u choose to help out on
something I can think your code will look better than mine.

And this also the main reason why I don't push code now...

I just need to get it done so I can retire...

Optimization and crap in the end. That's the stage where code should look
nice.

**Comment by KFlash** [comment]
@cherow/cherow-team I just pushed over 3100 tests to show off current state of Cherow 2.0. It is a fully working parser now on my computer, but still lack of a few things and the source code is a mess.

I'm still not ready to push any code!

**Comment by KFlash** [comment]
@cherow/cherow-team I pushed a almost finished statement code. Possible with feedback on it? Only a few more validations to fix plus fix the class decl code.

File: test/lexer/attachments.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import { State } from '../../src/state';
+import { Context } from '../../src/common';
+
+describe('Lexer - Attach comments', () => {
+
+  function pass(name: string, opts: any): void {
+      it(name, () => {
+          const state = new State(opts.source);
+          nextToken(state, Context.OptionsCollectComments);
+          t.deepEqual({
```
[the remark this answers is not in the document]
Comment (g-plane) [reply test/lexer/attachments.ts]: Just apply the changes directly.

File: src/types.ts [hunk 2]
```diff
@@ -1,33 +1,36 @@
...
 import { Comment } from './estree';
-import { CommentType } from './lexer/comments';
+import { Token } from './token';
 
 /**
- * `onToken` option.
+ * `ECMAScript version
  */
-export type OnToken = void | ((token: string, value: string, line?: number, column?: number) => void);
+export type EcmaVersion = 1 | 2 | 3 | 4 | 5 | 2015 | 2016 | 2017 | 2018 | 2019 | 2020;
```
[the remark this answers is not in the document]
Comment (g-plane) [reply src/types.ts]: Just apply the changes directly.

[the remark this answers is not in the document]
Comment (KFlash) [reply src/types.ts]: @Aladdin-ADD Did you change this?

File: src/parser/common.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+export function optional(state: ParserState, context: Context, token: Token): boolean {
+  if (state.currentToken !== token) return false;
+  nextToken(state, context);
+  return true;
+}
+
+export function expect(state: ParserState, context: Context, t: Token): boolean {
+  if (state.currentToken !== t) {
+      report(state, Errors.Unexpected);
+      return false;
```
[the remark this answers is not in the document]
Comment (fkleuver) [reply src/parser/common.ts]: In tolerant mode it doesn't throw right?

**Comment by KFlash** [comment]
You are right and that's code adopted from 1.6x. Tolerant mode not in 2.0.
I'm also working on 2.1 which will include the missing things in 2.0

So safe to remove

**Comment by g-plane** [comment]
I think there is no need to support IE.

**Comment by KFlash** [comment]
@g-plane Some old computers or servers still run legacy IE and what if they want to try to run Cherow 2.x? It will fail for them then and we got a few issue tickets to take care of. Is that what we want?

**Comment by fkleuver** [comment]
@KFlash Classes are a bit more efficient than objects. It's important that you eagerly define all properties in the constructor so that the browser doesn't need to generate multiple hidden classes for different "versions".

Most efficient however would be a hybrid between a class and a function scope since you don't have the member accessors. This will help performance if you need to store a bunch of information that might change a lot and doesn't necessarily all need to be accessible to the outside.

Can you give an example of what info you need to work with?

**Comment by KFlash** [comment]
@Aladdin-ADD It seems like that you *accidentaly* removed a state property `catch`. That one is to solve several edge cases in `try statement` while AnnexB is on / and off. You cant just remove that one!!!

**Comment by KFlash** [comment]
@fkleuver I'm aware of it. Just like TS does it, but it can tend to be slower.

**Comment by KFlash** [comment]
@fkleuver Did you open your gitter?

**Comment by fkleuver** [comment]
Problem with TS is that there is some lack of planning about how/when they assign and store variables. They're doing a lot of lazy initialization starting with `undefined` which makes it harder for the runtime to perform strongly typed optimizations. Same problem with the objects they generate, properties aren't assigned in a set order which leads to more hidden classes.
Other than that, TS performs pretty well and they do a lot of other stuff too. So I wouldn't say it's a bad example to look at for perf. There are still other things you're doing much more efficiently.

**Comment by fkleuver** [comment]
Yeah it seems I misunderstood what you meant. In the example you showed I don't think there is much difference in perf between objects or classes. As long as the created objects always have the same shape.
A naked class use 24 bits and a naked object uses 32 bits of memory however.. not that it matters much, but it's a funny little detail I discovered a few weeks ago

**Comment by KFlash** [comment]
Good to know you are back :) I'm still experimenting on stuff, but I'm reached a stage where the parser are running without validations.
The object literal parsing, any ideas for reducing that code a little more?

**Comment by fkleuver** [comment]
Hm before digging into optimizations, do you have a checklist somewhere with what's left to implement for the spec? How's the test coverage?

**Comment by fkleuver** [comment]
You merged the latest master as well right? Are you not getting a gazillion red squiggles locally with all the linting errors? Oo

**Comment by KFlash** [comment]
There is a reason why I haven't pushed any code yet, and that I actually wanted to keep the current 
Cherow 2.0 progress secret until i pushed something...

Anywyay...

**Parser code**

* Class expr / decl (*working on it just now!*)
* Regular expressions
* Template
* Regular expression validator

**Lexer**

* Complex identifiers included unicode
* Template
* Hex, octal and binary numbers


Test coverage? I guess around 95% on everything I have done so far.

**Comment by KFlash** [comment]
I started out with Master branch and keep this PR updated if Master changes

**Comment by fkleuver** [comment]
Lol, how come regular expressions always ends up on your todo list. I remember weeks and weeks of you working on just regular expressions.
So I'd say class expr / decl and template are the important ones here out of this list.
Circling back to object literals, if reducing code is what you're after then potentially the property definition parsing can be reused but I'm not sure what that would do to perf. I'd keep it separate for now..

What was wrong with the old regex code?

**Comment by KFlash** [comment]
The property definition parsing can be re-used with class expr / decl if you use a set of enums to define the origin and use a lot of if / else to control the parsing, but I'm sure it will end up in a lot of branching. And the fact that classes now also require a lot more like fields and soon decorators will make the list of if / else long and complex. So yeah I prefer to keep them separated for now, and eventually flatten it all into one huge recursive function later on if no perf drops.

Regex? I just enjoy playing with it :)

Classes is a WIP for me now. I'm done with the code and write tests

**Comment by KFlash** [comment]
@fkleuver What we actually need to think about is a new way of implement location tracking. The old one was kind of slow. Ideas?

**Comment by KFlash** [comment]
And I forgot to add export and import to the list. Getting that implementation correct isn't so easy if you want to do it correctly and also deal with scoping, lexical bindings, bound names etc. If you find any comfort in it. Acorn haven't implemented most of it except for lexicals. I expect to get a lot of gray hair.

**Comment by KFlash** [comment]
@fkleuver Maybe a good idea. Same thing as I did with the block scope implementation. It still require a lot of thinking

**Comment by fkleuver** [comment]
>  It still require a lot of thinking

Wouldn't be any fun otherwise would it ;p

**Comment by KFlash** [comment]
I'm open for all suggestions

File: src/types.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Aladdin-ADD) [reply src/types.ts]: no.

**Comment by Aladdin-ADD** [comment]
> @Aladdin-ADD It seems like that you accidentaly removed a state property catch. That one is to solve several edge cases in try statement while AnnexB is on / and off. You cant just remove that one!!!

what code? I just reformatted the code -- no changes but this commit: https://github.com/cherow/cherow/pull/252/commits/635405112bf48642655770b7174420ccee02475e

**Comment by KFlash** [comment]
@Aladdin-ADD It's all good. It's me that didn't publish that code. I thought I did

**Comment by KFlash** [comment]
@cherow/cherow-team Have a look at this new [parser.ts](https://github.com/cherow/cherow/blob/cherow_2/src/parser/parser.ts). This code is finale v. 2.0. 

All of you should have no problems to see the differences from 1.6.x and how complex it has become. 

@fkleuver I pass [this](https://github.com/cherow/cherow/blob/cherow_2/src/parser/parser.ts#L82) as function arguments instead of member access because I think it's way faster. Are you agree? And yes, this is something new in 2.0.

**Comment by KFlash** [comment]
@fkleuver If we now parse this one `export default function(){}; export default function(){};` we end up with this error message `SyntaxError: Line 1, column 57: Identifier '*default*' has already been declared`. You have a good suggestion for a error message?

And also for this cases

```js
export let a = 1, a = 2;
export const a = 1, a = 2;
```

**Comment by fkleuver** [comment]
Honestly I would go with what typescript says in those cases. They've got really good organization of error codes imo.
For declaring the same thing twice they say `Cannot redeclare (...) variable 'x'`
For the functions: `Cannot redeclare exported variable 'default'`
For the locals: `Cannot redeclare block-scoped variable 'a'`

Why not just stick to that?

**Comment by KFlash** [comment]
Sounds like a good idea. Do you have an link to this error messages?

File: src/parser/parser.ts [hunk 4]
```diff
@@ -17,80 +19,72 @@ export function parseSource(
...
+    // The flag to enable implied strict mode
+    if (options.impliedStrict) context |= Context.Strict;
+    // The flag to enable experimental features
+    if (options.experimental) context |= Context.OptionsExperimental;
+    // The flag to enable "native" NodeJS / V8 features
+    if (options.native) context |= Context.OptionsNative;
+    // The flag to enable tokenizing
+    if (options.tokenize) context |= Context.OptionsTokenize;
+    // The flag to disable web compability (AnnexB)
+    if (options.disableWebCompat) context &= ~Context.OptionDisablesWebCompat;
```
Comment (fkleuver) [inline src/parser/parser.ts]: Is this intentional?

Comment (KFlash) [reply src/parser/parser.ts]: This is an "hang over" from my experimental period.  By default we are parsing with web compability, so one of my experiments to remove AnnexB support. 

And yes, you are right. The code is wrong and will be fixed. I'm not using that code in the tests, only passing the context masks directly.

**Comment by fkleuver** [comment]
> Btw. You understand now what I actually have implemented into Cherow 2.0 that wasn't in 1.6x?

Yeah lol. You weren't tracking naming conflicts before. It never occurred to me before but it seems like a pretty crucial piece so it's nice to see it now.

I think using objects to track them like you're doing is probably solid.

**Comment by fkleuver** [comment]
These are the error messages: https://github.com/Microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json
And generated:
https://github.com/Microsoft/TypeScript/blob/master/lib/diagnosticMessages.generated.json

It's probably not a bad idea to keep them in a json like they do as well.

**Comment by KFlash** [comment]
@fkleuver Thanx. I will look into it.

One bad thing now is that I'm not sure I'm going to push any 2.0 code. It's way to complex atm., and I'm a afraid I can do a few mistakes resulting in a huge of issue tickets.

**Comment by fkleuver** [comment]
Why afraid? Nothing is forcing you to keep the code after pushing it, let alone releasing it to npm.

Which area are you most concerned about? Maybe I can help out at least with writing some tests. Tests are the savior when all else is uncertain :)

**Comment by KFlash** [comment]
First of all the code is a mess at the moment, but what I'm worried about is mostly the scoping and lexicals. I already got tons of tests backing it up, but I'm sure there are bugs somewhere.

**Comment by fkleuver** [comment]
There are a lot of ts compile errors when i pull `cherow_2` branch locally. I disabled a particular tslint rule which is just annoying (so go ahead and pull), but the ones that remain are kind of important.

For example:

image

It does not match up with the declared type, which in turn might lead you to assign values to the wrong var names:

image


If you want to store stuff in an object map, use the type `Record<string, whatever>`. Don't type them as `any` :P

**Comment by fkleuver** [comment]
Also how are you running tests? The whole thing completely and utterly fails for me locally, as well as on CI

**Comment by KFlash** [comment]
@fkleuver Hmm... you exposed pictures of code that wasn't aimed for public yet :) Please reconsider! But I will look into that soon.

The complexity is [this](https://github.com/cherow/cherow/blob/4f1520dc46230f499238e4b1578bfde5244f59ab/src/common.ts). Keeping the track of all this through the entire parsing

**Comment by KFlash** [comment]
Yes. Everything will fail for you of course because I haven't published the parser code yet. 90% of the code is missing :)

**Comment by KFlash** [comment]
@fkleuver Maybe you want to convert the masks to binary?

**Comment by KFlash** [comment]
@cherow/cherow-team This doesn't fail in acorn `var a; export {b as c};`. Should we report it?
and `export default [x] = y` doesn't parse in Esprima. Report this too?

**Comment by KFlash** [comment]
@cherow/cherow-team I just added almost done module code. [found here](https://github.com/cherow/cherow/blob/ea0e80bdd881f69daa88aef48c26d1b4569f772d/src/parser/module.ts). Any feedback on this one?

And the tests is [here for export](https://github.com/cherow/cherow/blob/cherow_2/test/parser/module/export.ts) and [here for import](https://github.com/cherow/cherow/blob/cherow_2/test/parser/module/import.ts)

Let me know if I missed any tests and / or you guys see something I fucked up :)

**Comment by KFlash** [comment]
I didn't push the `parseHoistableClassDeclaration` function and added tests for `class decl`. Else this is an 100% copy of what is running locale. And I also didn't add the code for decorators.

**Comment by KFlash** [comment]
@cherow/cherow-team **Important!!!!** We have a BC in 2.0 that also will affect the Acorn and Esprima parser. Cases like this

```js
export {x as default};
export { x as y };
export {x}
```

are valid in Acorn, but according to the specs they are not valid. They was also valid to parse in Cherow 1.6.x.  This is not valid because of the exported binding is not declared.

To get this valid, you need to parse something like this>

```js
let x = y; export {x as default};
let x = y; export { x as y };
let x = y; export {x}
```

**Comment by g-plane** [comment]
Can this feature be optional?

**Comment by KFlash** [comment]
Its not a feature. It just follow the ECMA specs

**Comment by KFlash** [comment]
@fkleuver See my latest commit. I added Test262 tests. They are not supposed to run yet, but can you clean up the files? They are running locale (*some of them*) so the setup is correct except for ....

**Comment by KFlash** [comment]
@cherow/cherow-team Ok, guys! The time have come that I'm now preparing the code locale for a push to this repo. It will take a few more days, and when I push, please not expect that things are perfect yet. But what I can promise is that what I push will surprise everyone ;)

**Comment by Aladdin-ADD** [comment]
@KFlash you have some uncommitted code? I've got a big number of errors and the tests not working.

**Comment by KFlash** [comment]
@Aladdin-ADD Yes, I have. I'm not ready to push all code yet. Still experimenting.

**Comment by KFlash** [comment]
@cherow/cherow-team 

I'm still experimenting and what you guys think about this alternative scanner code?

@fkleuver Any thougts on how to optimize this and is this the future?

```js
export function scan(source: string) {

  let index = 0;

  let start = 0;

  let line = 0;

  let column = 0;

  let length = source.length;

  let token = Token.EndOfSource;

  let value = '';

  let newline = false;

  let currentChar = source.charCodeAt(index);

  const unexpectedCharacter: () => void = () => report(index, line, column, Errors.Unexpected, String.fromCharCode(currentChar));

  const table = new Array(0xFFFF).fill(unexpectedCharacter, 0, 0x80) as((context: Context) => Token)[];

  table[Chars.Space] =
  table[Chars.Tab] =
  table[Chars.FormFeed] =
  table[Chars.VerticalTab] = () => {
    return Token.WhiteSpace;
  }

  // `a`...z`
  for (let i = Chars.LowerA; i <= Chars.LowerZ; i++) table[i] = scanIdentifier;

  function scanIdentifier(): any {
    value = source.slice(start, index);
    return Token.Identifier;
  }

  return function(context: Context): any {

  while (index < length) {
        let current = currentChar;
        start = index;
        currentChar = source.charCodeAt(++index);
        if (((token = table[current](context) & Token.WhiteSpace)) !== Token.WhiteSpace) {
          return {
            type: token,
            value,
            newline,
            start,
            end: index
          };
        }
    }

    return {
      type: token,
      value: '',
      newline,
      start,
      end: index
    };
  };
}
```

**Comment by KFlash** [comment]
@fkleuver If you are ok with the new lexer code, can you chat me on gitter. I have some code to discuss that I don't go online with yet. Need ideas!

**Comment by KFlash** [comment]
@cherow/cherow-team Alternative scanner code can be [located here](https://github.com/cherow/cherow/blob/cherow_2/src/lexer_demo.ts)
Thoughts?

**Comment by KFlash** [comment]
@fkleuver Did you study the alternative lexer code? Should we use that one or use what we have now? The demo code is way faster I guess - at least when it comes to WS skipping.

**Comment by KFlash** [comment]
@fkleuver If you look at the alternative lexer code you will find out that use of the tables kill the performance. With both the `table` and the `whitespaceTable` it will use `1 second` to perform `115 tests`.
If you replace the table with a switch statement - you will get this `115 passing (40ms)`

What do you think is the reason for this? Anyway. The alternative lexer code seems to perform better than the current one if we use switch statements.

**Comment by KFlash** [comment]
@cherow/cherow-team Updated the demo to show off how the parser code will look like with alternative scanner code. [Seen here](https://github.com/cherow/cherow/blob/cherow_2/src/lexer_demo.ts#L913)

For now Cherow 2.0 is put on hold until I know wich path to take, and @fkleuver thougs about it

**Comment by fkleuver** [comment]
@KFlash I see you went with the function-scope variables approach we talked about earlier. In combination with that, you also reverted from table to very large switch.
The move away from table is necessary for the function scope thing unfortunately, because I'm still quite sure a table would perform better than a very large switch. I have done countless benchmarks proving that array indexers outperform switches with a tipping point at around 7-8 cases.
If there were a way to have both, performance would be even better, but that might not be possible or practical.

I do believe the scanner can be further optimized by doing some hot path analysis on which ones are hit most often, moving them up in separate if/else statements, and certainly eliminating the a-z and A-Z cases from [here](https://github.com/cherow/cherow/blob/cherow_2/src/lexer_demo.ts#L819) to [https://github.com/cherow/cherow/blob/cherow_2/src/lexer_demo.ts#L872](here) and replacing it with a range check like this:
```ts
if ((ch > 64 && ch < 91) || (ch > 96 && ch < 123)) {
  scanKnownIdentifier()
}
```

Remember that a `switch` statement is translated to a binary jump table, so with 50+ cases like that you'll usually have at least 4 jumps multiplied by anywhere between 1-4 `if` branches, giving you between 4 and 16 `if` checks for what could be normalized to a single `if` check with at most 4 branches.

Try to keep switches no larger than 7-8 items. If they do get larger, some table or range trick might help a bit.

**Comment by KFlash** [comment]
@fkleuver Read my gitter response to you. And then you will understand how I'm thinking further.
Regarding the lexer code. I have to get back to it later on if you don't want to take a stab at it and try to optimize it? I only optimized for whitespace skipping and I know there are space for improvements. Also the string literal code both has a bug and need to be optimized. E.g. Returning the Token directly when you find the closing quote in the iteration loop will increase performance with 2%

**Comment by KFlash** [comment]
@cherow/cherow-team 

This is the current benchmark result for Cherow 2.0. It's almost complete at this stage when I ran the benchmark. **But** this is with prototype. If I try do a singleton module with function scope the performance get so horrible that I'm afraid to show you guys the results here.

Further. The benchmark below is with full lexical scoping in the background - meaning it should be a real slow down if we look at Acorn. But as seen in the benchmark that is not the case. Except for *let**.

I think there are still room for improvements, and this is soon going to land now. But I need to finish a few more things first. But soon..... I said a beta will land in january 2019 and I'm ahead of my schedule.

bench

**Comment by KFlash** [comment]
@cherow/cherow-team ` 53 passing (16ms)` for punctuator scanning alone means the scanner code is extreme fast. With the demo the average was 70ms and current Cherow 1.0 is 70 - 90 ms in average.

**Comment by KFlash** [comment]
@cherow/cherow-team I'm closing this. It become too big and complex. A beta release PR will land later on.

**Pull request closed without merging**